hibernate-core/src/main/java/org/hibernate/id/enhanced/PooledLoThreadLocalOptimizer.java

```diff
@@ -127,4 +122,10 @@ public IntegralDataTypeHolder getLastSourceValue() {
 	public boolean applyIncrementSizeToSourceValues() {
 		return true;
 	}
+
+	private static class GenerationStates {
```

Comment:
I'm thinking that this implementation is not a very good choice for people who are using multi-tenancy, when it's combined with "many tenants" and many threads, as the state gets rather large, with significant ID fragmentation on the DB level.

So while it's of course a choice we want to allow as the definition of such trade-offs is subjective, I would expect this Optimizer being used primarily by people who are not using multitenancy; but this wrapper object penalises them as it leads to more memory waste by repeating this structure for each thread.

Could be simple to use two different ThreadLocals, one holding `GenerationState` for single-tenant usage, and one holding the map for multi-tenant?

Reply:
```
I'm thinking that this implementation is not a very good choice for people who are using multi-tenancy, when it's combined with "many tenants" and many threads, as the state gets rather large, with significant ID fragmentation on the DB level.
```

It's a decent choice if you only have a few tenants, and a bad one if you have thousands of them.
Regardless, this is how this optimizer is supposed to work, as described in the documentation. People who don't want this state to be held in a thread-local can simply use the `pooled-lo` optimizer instead of the `pooled-lotl` optimizer.

```
I would expect this Optimizer being used primarily by people who are not using multitenancy; but this wrapper object penalises them as it leads to more memory waste by repeating this structure for each thread.
```

The map is `null` initially, and stays `null` as long as you don't use multi-tenancy. So people relying on single-tenancy are only penalized by the extra `GenerationStates` instance, which doesn't seem so bad?

Anyway, I'll do that if you feel strongly about this.

Reply:
Done.

Reply:
thanks, and sorry for the nitpicking.

FWIW I wasn't feeling strongly about it, just a small suggestion but yes I think it's slightly better - even for null values, the wrapping object still needs to be allocated (and null values still need space for the pointer, even if it's null).